src/cluster_legacy.c

```diff
@@ -4323,22 +4362,50 @@ void clusterSendUpdate(clusterLink *link, clusterNode *node) {
  *
  * If link is NULL, then the message is broadcasted to the whole cluster. */
 void clusterSendModule(clusterLink *link, uint64_t module_id, uint8_t type, const char *payload, uint32_t len) {
-    uint32_t msglen = sizeof(clusterMsg) - sizeof(union clusterMsgData);
-    msglen += sizeof(clusterMsgModule) - 3 + len;
-    clusterMsgSendBlock *msgblock = createClusterMsgSendBlock(CLUSTERMSG_TYPE_MODULE, msglen);
-
-    clusterMsg *hdr = getMessageFromSendBlock(msgblock);
-    hdr->data.module.msg.module_id = module_id; /* Already endian adjusted. */
-    hdr->data.module.msg.type = type;
-    hdr->data.module.msg.len = htonl(len);
-    memcpy(hdr->data.module.msg.bulk_data, payload, len);
-
-    if (link)
-        clusterSendMessage(link, msgblock);
-    else
-        clusterBroadcastMessage(msgblock);
+    clusterMsgSendBlock *msgblock = NULL, *msgblock_light = NULL;
+    ClusterNodeIterator iter;
 
-    clusterMsgSendBlockDecrRefCount(msgblock);
+    if (link) {
+        clusterNodeIterNode(&iter, link->node);
+    } else {
+        /* Broadcast to all the nodes. */
+        clusterNodeIterInitAllNodes(&iter);
+    }
+    clusterNode *node;
+    while ((node = clusterNodeIterNext(&iter)) != NULL) {
+        if (node->flags & (CLUSTER_NODE_MYSELF | CLUSTER_NODE_HANDSHAKE)) continue;
+        if (nodeSupportsLightMsgHdrForModule(node)) {
+            if (msgblock_light == NULL) {
+                uint32_t msglen_light = sizeof(clusterMsgLight) - sizeof(union clusterMsgData);
+                msglen_light += sizeof(clusterMsgModule) - 3 + len;
+                int msgtype_light = CLUSTERMSG_TYPE_MODULE;
+                msgtype_light |= CLUSTERMSG_LIGHT;
+                msgblock_light = createClusterMsgSendBlock(msgtype_light, msglen_light);
+                clusterMsgLight *hdr = getLightMessageFromSendBlock(msgblock_light);
+                hdr->data.module.msg.module_id = module_id; /* Already endian adjusted. */
+                hdr->data.module.msg.type = type;
+                hdr->data.module.msg.len = htonl(len);
+                memcpy(hdr->data.module.msg.bulk_data, payload, len);
+            }
+            clusterSendMessage(node->link, msgblock_light);
```

Comment:
There is some duplicated code for light and heavy here, right?

This function corresponds to clusterPropagatePublish (for module instead of publish), but this is more code compared to the publish one. I can see that clusterPropagatePublish calls a helper function clusterCreatePublishMsgBlock with an `is_light` flag. It seems to reduce the duplicated code a bit. Maybe do a similar refactoring here?